### PR TITLE
Add ancestor, precedingSibling and followingSibling M1 extension methods

### PR DESCRIPTION
**Pull request by erk-dik**
This PR includes the following M1 extension methods (mentioned in issue #21) and their tests.

- Ancestor
- PrecedingSiblings
- FollowingSiblings

### CODE REVIEW

File: src/LionWeb-CSharp/core/M1/M1Extensions.cs [hunk 1]
```diff
@@ -87,6 +87,45 @@ public static void InsertAfter(this INode self, INode newSuccessor)
         list.Insert(index + 1, newSuccessor);
         parent.Set(containment, list);
     }
+    
+    //TODO: The order of returned siblings ? 0 ... n or n ...0
+    public static IEnumerable<INode> PrecedingSibling(this INode self, bool includeSelf = false)
+    {
+        INode? parent = self.GetParent();
```
Comment (enikao) [inline src/LionWeb-CSharp/core/M1/M1Extensions.cs:94]: We now have the same code in 4 methods, definitely time to factor it out.

File: src/LionWeb-CSharp/core/M1/M1Extensions.cs [hunk 2]
```diff
@@ -87,6 +87,45 @@ public static void InsertAfter(this INode self, INode newSuccessor)
...
+        if (!containment.Multiple)
+            throw new TreeShapeException(self, "A single containment does not have siblings");
+        
+        var value = parent.Get(containment);
+        if (value is not IEnumerable enumerable)
+            // should not happen
+            throw new TreeShapeException(self, "A single containment does not have siblings");
+        
+        var list = new List<INode>(containment.AsNodes<INode>(enumerable));
+        return includeSelf ? list.Take(list.IndexOf(self) + 1) : list.Take(list.IndexOf(self));
```
Comment (enikao) [inline src/LionWeb-CSharp/core/M1/M1Extensions.cs:108]: `includeSelf` only affects the parameter to `Take()`, so move the condition there.

File: src/LionWeb-CSharp/core/M1/M1Extensions.cs (hunk 1 above)
Comment (erk-dik) [reply src/LionWeb-CSharp/core/M1/M1Extensions.cs]: Repeated code factored out forPrecedingSiblings and FollowingSibliblings, but not for InsertAfter and InsertBefore, the reasons are the following:

-  Factored out method return a list of nodes to forPrecedingSiblings and FollowingSibliblings, but InsertAfter and InsertBefore makes use of parent, containment variables after the non-repeated code block.
- exception messages are different for PrecedingSiblings, FollowingSibliblings and  InsertAfter and InsertBefore. Of course those could be parameters to the factored out method.

File: test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs [hunk 3]
```diff
@@ -138,6 +138,164 @@ public void ManyParents_Annotation_Self()
...
+    
+    [TestMethod]
+    public void Ancestor_ManyParents()
+    {
+        var node = new Coord("n");
+        var parent = new Line("p") { Start = node };
+        var ancestor = new Geometry("a") { Shapes = [parent] };
+        Assert.IsNotNull(node.Ancestor<Shape>());
+        Assert.AreEqual(parent, node.Ancestor<Shape>());
+    }
```
Comment (JojoEffect) [inline test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs:167]: Should the `ancestor` (`Geometry`) be tested here too?

File: test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs [hunk 4]
```diff
@@ -138,6 +138,164 @@ public void ManyParents_Annotation_Self()
...
+    
+    [TestMethod]
+    public void Ancestor_ManyParents_Self()
+    {
+        var node = new Coord("n");
+        var parent = new Line("p") { Start = node };
+        var ancestor = new Geometry("a") { Shapes = [parent] };
+        Assert.IsNotNull(node.Ancestor<Shape>(true));
+        Assert.AreEqual(parent, node.Ancestor<Shape>(true));
+    }
```
Comment (JojoEffect) [inline test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs:177]: Should the `ancestor` (`Geometry`) be tested here too?

File: src/LionWeb-CSharp/core/M1/M1Extensions.cs [hunk 5]
```diff
@@ -164,6 +215,24 @@ public static void InsertAfter(this INode self, INode newSuccessor)
     public static IEnumerable<INode> Ancestors(this INode self, bool includeSelf = false) =>
         Ancestors<INode>(self, includeSelf);
```
Comment (JojoEffect) [inline src/LionWeb-CSharp/core/M1/M1Extensions.cs:217]: @enikao lets have a short talk about this exceptions being thrown here if the given node does not fulfill all reuirements. I'm not sure if a user expects this (even if it is documented at the API)

File: test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs
Comment (JojoEffect) [inline test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs:1]: Is this filename a good one? Maybe I'm just not used to the term "Axis" in the context of models/trees/graphs!?

Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs:1]: Niko would better comment on this, but to my knowledge the term "axis" comes from XML path language (XPath) specs : https://www.w3.org/TR/1999/REC-xpath-19991116/#axes

Comment (enikao) [reply test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs:1]: Erkan quotes the right source; we could use the correct plural "Axes".

File: test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs (hunk 4 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs:177]: Added.

File: test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs:1]: @enikao shall I rename the test file ? ( AxisTests.cs -> AxesTests.cs )

Comment (JojoEffect) [reply test/LionWeb-CSharp-Test/tests/nodeapi/AxisTests.cs:1]: Suggestion from @enikao and me: "TreeTraversalTests" or "TTT" or "T³" (last both are just a joke)

**Review by JojoEffect — APPROVED** [review]

File: src/LionWeb-CSharp/core/M1/M1Extensions.cs [hunk 6]
```diff
@@ -87,6 +87,43 @@ public static void InsertAfter(this INode self, INode newSuccessor)
...
+    /// Optionally includes <paramref name="self"/>.
+    /// 
+    /// Starting from index 0 to the index of base node, includes the index of <param name="self"></param>
+    /// if <paramref name="includeSelf"/> is set to true.  
+    /// </summary>
+    /// <param name="self">Base node to find preceding siblings of.</param>
+    /// <param name="includeSelf">If true, the result includes <paramref name="self"/>.</param>
+    /// <returns>List of preceding siblings of <param name="self"></param>.</returns>
+    /// <exception cref="TreeShapeException">If <paramref name="self"/> has no parent, or is not a child in a multiple containment.</exception>
+    public static IEnumerable<INode> PrecedingSibling(this INode self, bool includeSelf = false)
```
Comment (enikao) [inline src/LionWeb-CSharp/core/M1/M1Extensions.cs:102]: We return more than one, so name should be plural: `PrecedingSiblings`

File: src/LionWeb-CSharp/core/M1/M1Extensions.cs [hunk 7]
```diff
@@ -87,6 +87,43 @@ public static void InsertAfter(this INode self, INode newSuccessor)
...
+    ///
+    /// Starting from the following index of <param name="self"></param>, includes the index of <paramref name="self"/>
+    /// if <paramref name="includeSelf"/> is set to true.  
+    /// </summary>
+    ///
+    /// <param name="self">Base node to find following siblings of.</param>
+    /// <param name="includeSelf">If true, the result includes <paramref name="self"/>.</param>
+    /// <returns>List of following siblings of <param name="self"></param>.</returns>
+    /// <exception cref="TreeShapeException">If <paramref name="self"/> has no parent, or is not a child in a multiple containment.</exception>
+    public static IEnumerable<INode> FollowingSibling(this INode self, bool includeSelf = false)
```
Comment (enikao) [inline src/LionWeb-CSharp/core/M1/M1Extensions.cs:121]: plural

File: src/LionWeb-CSharp/core/M1/M1Extensions.cs [hunk 8]
```diff
@@ -154,6 +191,20 @@ public static void InsertAfter(this INode self, INode newSuccessor)
         bool includeAnnotations = false) =>
         Children<INode>(self, includeSelf, includeAnnotations);
 
+    
+    /// <summary>
+    /// Enumerates all direct and indirect parents of <paramref name="self"/> that matches with type <typeparamref name="T"/>
```
Comment (enikao) [inline src/LionWeb-CSharp/core/M1/M1Extensions.cs:196]: Single return, so no enumeration

File: src/LionWeb-CSharp/core/M1/M1Extensions.cs [hunk 9]
```diff
@@ -154,6 +191,20 @@ public static void InsertAfter(this INode self, INode newSuccessor)
...
+    /// Optionally includes <paramref name="self"/>.
+    /// </summary>
+    /// <param name="self">Base node to find ancestor of.</param>
+    /// <param name="includeSelf">If true, the result includes <paramref name="self"/>.</param>
+    /// <typeparam name="T"></typeparam>
+    /// <returns>The first ancestor of <paramref name="self"/> that matches with type <typeparamref name="T"/> or default </returns>
```
Comment (enikao) [inline src/LionWeb-CSharp/core/M1/M1Extensions.cs:202]: use `<c>null</c>` instead of "default", way easier to understand (and always correct)

**Review by enikao — APPROVED** [review]

**Pull request merged**